src/daft-local-execution/src/intermediate_ops/intermediate_op.rs

```diff
@@ -3,9 +3,132 @@ use std::sync::Arc;
+
+/// The number of rows that will trigger an intermediate operator to output its data.
+pub const OUTPUT_THRESHOLD: usize = 1000;
+
+/// State of an operator task, used to buffer data and output it when a threshold is reached.
+pub struct OperatorTaskState {
+    pub buffer: Vec<Arc<MicroPartition>>,
+    pub curr_len: usize,
+    pub threshold: usize,
+}
+
+impl OperatorTaskState {
+    pub fn new() -> Self {
+        Self {
+            buffer: vec![],
+            curr_len: 0,
+            threshold: OUTPUT_THRESHOLD,
+        }
+    }
+
+    pub fn add(&mut self, part: Arc<MicroPartition>) {
+        self.buffer.push(part);
+        self.curr_len += 1;
+    }
+
+    pub fn try_clear(&mut self) -> Option<DaftResult<Arc<MicroPartition>>> {
+        if self.curr_len >= self.threshold {
+            self.clear()
+        } else {
+            None
+        }
+    }
+
+    pub fn clear(&mut self) -> Option<DaftResult<Arc<MicroPartition>>> {
+        if self.buffer.is_empty() {
+            return None;
+        }
+        let concated =
+            MicroPartition::concat(&self.buffer.iter().map(|x| x.as_ref()).collect::<Vec<_>>())
+                .map(Arc::new);
+        self.buffer.clear();
+        self.curr_len = 0;
+        Some(concated)
+    }
+}
+
+pub struct IntermediateOpActor {
+    sender: MultiSender,
+    receiver: MultiReceiver,
+    op: Box<dyn IntermediateOperator>,
+}
+
+impl IntermediateOpActor {
+    pub fn new(
+        op: Box<dyn IntermediateOperator>,
+        receiver: MultiReceiver,
+        sender: MultiSender,
+    ) -> Self {
+        Self {
+            op,
+            receiver,
+            sender,
+        }
+    }
+
+    async fn single_operator_task(
+        mut receiver: SingleReceiver,
+        sender: SingleSender,
+        op: Box<dyn IntermediateOperator>,
+    ) -> DaftResult<()> {
+        let mut state = OperatorTaskState::new();
+        while let Some(morsel) = receiver.recv().await {
+            let result = op.execute(&morsel?)?;
+            state.add(result);
+            if let Some(part) = state.try_clear() {
+                let _ = sender.send(part).await;
+            }
+        }
+        if let Some(part) = state.clear() {
+            let _ = sender.send(part).await;
+        }
+        Ok(())
+    }
+
+    pub async fn run(&mut self) -> DaftResult<()> {
+        let mut inner_task_senders: Vec<SingleSender> =
+            Vec::with_capacity(self.sender.buffer_size());
+        let mut curr_sender_idx = 0;
+
+        while let Some(morsel) = self.receiver.recv().await {
+            if let Some(s) = inner_task_senders.get(curr_sender_idx) {
+                let _ = s.send(morsel).await;
+            } else {
+                let next_sender = self.sender.get_next_sender();
+                let (single_sender, single_receiver) = create_single_channel(1);
+
+                let op = self.op.clone();
+                tokio::spawn(async move {
```

Comment:
we shouldn't have any `tokio::spawn` in our code. have a util function that is called
`spawn_compute(...)`